Display lesson action blocks in the frontend

gkaragia opened the pull request:
### Changes proposed in this Pull Request

* This PR handles frontend rendering of the lesson actions block.
* The old template deprecation will happen as part of another PR so the old buttons/notices are still displayed.
* The rules of when each button will appear are the following:
  * The 'Next Lesson' button will appear when the course is completed.
  * The 'Reset Lesson' button will appear when the user has completed the course, there is a quiz linked to it and the quiz reset button is enabled in both block and quiz settings.
  * The 'Complete Lesson' button will appear when the user has not completed the course and there is no required quiz to the lesson.
  * The 'View Quiz' button will appear when there is a quiz linked to the course and the lesson is not completed.

### Testing instructions

* For the described cases, observe that the buttons are displayed correctly in the frontend.
* Try various settings on the blocks and observe that they are displayed correctly in the frontend.

gkaragia commented on assets/blocks/lesson-actions/complete-lesson-block/index.js:
```diff
@@ -12,7 +12,7 @@ export default createButtonBlockType( {
 		title: __( 'Complete lesson', 'sensei-lms' ),
 		description: __(
-			'Enable an enrolled user to mark the lesson as complete.',
+			'Enable an enrolled user to mark the lesson as complete. The button is displayed when the user is enrolled and there is no required quiz linked to the lesson.',
```
@donnapep I updated the descriptions of the lesson buttons since I think that we should describe to the user when each of the buttons is going to be displayed in the frontend. Let me know of any suggestions.

gkaragia commented on includes/blocks/class-sensei-lesson-actions-block.php:
```diff
@@ -37,6 +37,16 @@ public function __construct() {
+			return '';
+		}
+
+		if ( ! Sensei_Lesson::should_show_lesson_actions( $lesson->ID ) ) {
```
I took most of the checks from the old buttons. If we wan't to update/streamline the rules of when each button is going to be displayed, this PR is a good chance to do it.

yscik commented on includes/blocks/class-sensei-reset-lesson-block.php:
```diff
@@ -0,0 +1,81 @@
+}
+
+/**
+ * Class Sensei_Complete_Lesson_Block is responsible for rendering the 'Complete Lesson' block.
```
Comments here need to be updated.

gkaragia replied:
Fixed in d6afb43efa324c5c58e005af615dfbf003fe3a89

yscik reviewed (commented):
I'm wondering if we should show the `Reset lesson` button even when there is no quiz, since it also resets the lesson completed state. 
If not, we should update the editor too, to only show it when there is a quiz.

gkaragia commented on assets/blocks/lesson-actions/complete-lesson-block/index.js:
```diff
@@ -25,6 +25,9 @@ export default createButtonBlockType( {
 				default: __( 'Complete Lesson', 'sensei-lms' ),
 			},
+			buttonClassName: {
+				default: [ 'sensei-stop-double-submission' ],
```
> This should not be added based on an editable block attribute, since it's a technical class that always needs to be present. The rendering logic in PHP would be a better place for it.

I would like to avoid regexp replacements if possible. Ideally this should be placed in the save function but the complexity is that the button is shared and we only want to add it in cases where the button is part of form.

yscik replied:
We could think about moving the buttons to be fully PHP-rendered based on the attributes, instead of the current regexp approach ([class-sensei-block-take-course.php#L93](https://github.com/Automattic/sensei/blob/e65e0d539203a6c833bcc9737f6d2e596d205589/includes/blocks/class-sensei-block-take-course.php#L93)) for changing the HTML generated by the save function.

gkaragia replied:
> We could think about moving the buttons to be fully PHP-rendered based on the attributes

We could but won't this be duplicated work? We used a regexp previously as the block was already released and we would need versioning. I also think that being able to add a class to the button is a useful attribute and it could be also used in the future.

alexsanford commented:
> The 'Reset Lesson' button will appear when the user has completed the course, there is a quiz linked to it and the quiz reset button is enabled in both block and quiz settings.

This seems pretty confusing to me.

- Making the "Reset Lesson" button visible even without a Quiz available seems sensible, as suggested above.
- I wonder if we can move away from needing to enable this in two places, somehow? Or at least indicate in the UI when the button will not be shown, like we do for "View Quiz". The problem with that is it isn't obvious that it is enabled by the "Allow user to retake the quiz" setting, so the user is unlikely to know how to enable it.

alexsanford commented:
There's an edge case we may have missed here. When a quiz is completed, currently Sensei keeps the "View the Lesson Quiz" button on the page, so you can view the completed quiz. Similarly, when a quiz is optional, if you complete the Lesson, there is still the option to go back later and view or take the quiz.

After this PR (and the [next one](https://github.com/Automattic/sensei/pull/3935) to remove the old buttons), we will no longer have the ability to review the quiz (or take the optional quiz) after the Lesson has been completed.

gkaragia commented:
> There's an edge case we may have missed here. When a quiz is completed, currently Sensei keeps the "View the Lesson Quiz" button on the page, so you can view the completed quiz. Similarly, when a quiz is optional, if you complete the Lesson, there is still the option to go back later and view or take the quiz.
>
> After this PR (and the next one to remove the old buttons), we will no longer have the ability to review the quiz (or take the optional quiz) after the Lesson has been completed.

Yes this was discussed earlier we decided to not show the button when the course is completed. The problem is that if we show the 'View Quiz' button when the course is completed then there is a case that we would display three buttons: 'Next lesson', 'Reset Lesson' and 'View Quiz'. Currently there is no hint that three buttons can be displayed in the editor. Also if we plan to display three buttons it would make more sense to have them in the same line.

gkaragia commented:
> I wonder if we can move away from needing to enable this in two places, somehow? Or at least indicate in the UI when the button will not be shown, like we do for "View Quiz". The problem with that is it isn't obvious that it is enabled by the "Allow user to retake the quiz" setting, so the user is unlikely to know how to enable it.

Yep we could do the same with 'View Quiz' button. Should we also consider dropping the block setting which shows or hides the Reset button? I am thinking that the block setting should always follow the quiz setting as every other combination is not useful. Or is there another case which I am missing?

gkaragia commented:
I added the Reset lesson button in lessons with no quiz in commit 3c003dfc57a13ba42c226cd47551eb42107f6c6e. I also removed the 'Next Lesson' button from the notice if the action block exists in 7f4120de349807d6bcd10265ba2367191f3dd7a1.

yscik reviewed (dismissed):
LGTM 🚢

gkaragia commented:
Hey @alexsanford proceeded with merging this one. Let me know if we want to raise a new issue to handle what is mentioned here https://github.com/Automattic/sensei/pull/3929#issuecomment-768881845

yscik commented:
Hm, just thought about it a bit more: what if we always display the reset button it the lesson is completed, and reset either just the lesson progress, or both the lesson progress and quiz result if that's enabled. Though does it even make sense have the lesson in an uncompleted state while still having quiz results?

gkaragia commented:
> Hm, just thought about it a bit more: what if we always display the reset button it the lesson is completed, and reset either just the lesson progress, or both the lesson progress and quiz result if that's enabled. 

I had the impression that this was the plan and it is how it works currently. If it doesn't there is a bug 😄 

> Though does it even make sense have the lesson in an uncompleted state while still having quiz results?

If I understand the question correctly there is a case for the lesson to be not completed while there are quiz results. It can happen if the user failed or if there is a pending grading.

yscik commented:
`Reset Lesson` is not rendering now if there is a quiz with quiz reset disabled. ([condition](https://github.com/Automattic/sensei/blob/3c003dfc57a13ba42c226cd47551eb42107f6c6e/includes/blocks/class-sensei-reset-lesson-block.php#L55)) 

Also looks like lesson reset (`sensei_remove_user_from_lesson`) always removes quiz answers: https://github.com/Automattic/sensei/blob/7f4120de349807d6bcd10265ba2367191f3dd7a1/includes/class-sensei-utils.php#L709

gkaragia commented:
> `Reset Lesson` is not rendering now if there is a quiz with quiz reset disabled. ([condition](https://github.com/Automattic/sensei/blob/3c003dfc57a13ba42c226cd47551eb42107f6c6e/includes/blocks/class-sensei-reset-lesson-block.php#L55))

Ah it seems that I got the earlier suggestion wrong. So the current functionality is to show the 'Reset' button only when the course is completed. Not sure if we should move away from that and allow to always clear the lesson status mainly because I am not sure what effect would have to delete statuses like 'ungraded' or 'graded'.

donnapep commented:
> Yes this was discussed earlier and we decided to not show the button when the course is completed.

We absolutely need to show the _View Quiz_ button if the course is completed. Students need to be able to view their quiz after it's been graded. If there are UI challenges with this, then I'm afraid we'll need to figure those out.

For the _Reset Lesson_ button, it seems that there may be a lot of edge cases we need to figure out if we change the rules around its visibility. Needing to make more decisions slows us down. Given the beta is supposed to go out today, I think it may be best to keep things as they are in the current version, since things we hadn't considered are likely to come up either now or later.
